daml-lf/upgrades/src/main/scala/com/digitalasset/daml/lf/upgrades/Typecheck.scala

```diff
@@ -0,0 +1,235 @@
+  def fold[B](f: (A, A) => B): B = f(past, present)
+  def zip[B, C](that: Upgrading[B], f: (A, B) => C) =
+    Upgrading(f(this.past, that.past), f(this.present, that.present))
+}
+
+case class UpgradeError(message: String) extends Throwable(message)
+
+sealed abstract class UpgradedRecordOrigin
+
+final case class TemplateBody(template: Ref.DottedName) extends UpgradedRecordOrigin
+final case class TemplateChoiceInput(template: Ref.DottedName, choice: Ref.ChoiceName)
+    extends UpgradedRecordOrigin
+final case object TopLevel extends UpgradedRecordOrigin
+
+object Typecheck {
+  def typecheckUpgrades(
+      present: (Ref.PackageId, Ast.Package),
+      mbPast: Option[(Ref.PackageId, Ast.Package)],
+  ): Try[Unit] = {
+    mbPast match {
+      case None => Failure(UpgradeError("CouldNotResolveUpgradedPackageId"));
+      case Some(past) => {
+        val tc = this(Upgrading(past, present))
+        tc.check()
+      }
+    }
+  }
+}
+
+case class Typecheck(packagesAndIds: Upgrading[(Ref.PackageId, Ast.Package)]) {
+  lazy val packageId: Upgrading[Ref.PackageId] = packagesAndIds.map(_._1)
+  lazy val _package: Upgrading[Ast.Package] = packagesAndIds.map(_._2)
+
+  def extractDelExistNew[K, V](
+      arg: Upgrading[Map[K, V]]
+  ): (Map[K, V], Map[K, Upgrading[V]], Map[K, V]) =
+    (
+      arg.past -- arg.present.keySet,
+      arg.past.keySet
+        .intersect(arg.present.keySet)
+        .map(k => k -> Upgrading(arg.past(k), arg.present(k)))
+        .toMap,
+      arg.present -- arg.past.keySet,
+    )
+
+  def checkDeleted[K, V, T <: Throwable](
+      arg: Upgrading[Map[K, V]],
+      handler: V => T,
+  ): Try[(Map[K, Upgrading[V]], Map[K, V])] = {
+    val (deletedV, existingV, newV) = extractDelExistNew(arg)
+    deletedV.headOption match {
+      case Some((k @ _, v)) => Failure(handler(v))
+      case _ => Success((existingV, newV))
+    }
+  }
+
+  def tryAll[A, B](t: Iterable[A], f: A => Try[B]): Try[Seq[B]] =
+    Try { t.map(f(_).get).toSeq }
+
+  def check(): Try[Unit] = {
+    for {
+      (upgradedModules, newModules @ _) <-
+        checkDeleted(
+          _package.map(_.modules),
+          (m: Ast.Module) => UpgradeError(s"MissingModule(${m.name})"),
+        )
+      _ <- tryAll(upgradedModules.values, checkModule(_))
+    } yield ()
+  }
+
+  def checkModule(module: Upgrading[Ast.Module]): Try[Unit] = {
+    def datatypes(module: Ast.Module): Map[Ref.DottedName, Ast.DDataType] =
+      module.definitions.flatMap(_ match {
+        case (k, v: Ast.DDataType) => Some((k, v));
+        case _ => None;
+      })
+
+    for {
+      (existingTemplates, _new) <- checkDeleted(
+        module.map(_.templates),
+        (_: Ast.Template) => UpgradeError(s"MissingTemplate(t)"),
+      )
+      _ <- tryAll(existingTemplates.values, checkTemplate(_))
+
+      (existingDatatypes, _new) <- checkDeleted(
+        module.map(datatypes(_)),
+        (_: Ast.DDataType) => UpgradeError(s"MissingDataCon(t)"),
```

Comment:
I'd include the actual error subject if possible for all the the error cases. Not sure if that was your aim with with the string interpolators (`s"..."`). In any case, if there's nothing to interpolate, I'd remove the `s`